pkg/limits/ingest_limits.go

```diff
@@ -0,0 +1,669 @@
+	// Stream metadata older than WindowSize will be evicted from the metadata map.
+	WindowSize time.Duration `yaml:"window_size"`
+
+	// LifecyclerConfig is the config to build a ring lifecycler.
+	LifecyclerConfig ring.LifecyclerConfig `yaml:"lifecycler,omitempty"`
+
+	KafkaConfig kafka.Config `yaml:"-"`
+
+	// The number of partitions for the Kafka topic used to read and write stream metadata.
+	// It is fixed, not a maximum.
+	NumPartitions int `yaml:"num_partitions"`
+}
+
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.LifecyclerConfig.RegisterFlagsWithPrefix("ingest-limits.", f, util_log.Logger)
+	f.BoolVar(&cfg.Enabled, "ingest-limits.enabled", false, "Enable the ingest limits service.")
+	f.DurationVar(&cfg.WindowSize, "ingest-limits.window-size", 1*time.Hour, "The time window for which stream metadata is considered active.")
+	f.IntVar(&cfg.NumPartitions, "ingest-limits.num-partitions", 64, "The number of partitions for the Kafka topic used to read and write stream metadata. It is fixed, not a maximum.")
+}
+
+type metrics struct {
+	tenantStreamEvictionsTotal *prometheus.CounterVec
+
+	kafkaReadLatency    prometheus.Histogram
+	kafkaReadBytesTotal prometheus.Counter
+}
+
+func newMetrics(reg prometheus.Registerer) *metrics {
+	return &metrics{
+		tenantStreamEvictionsTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Namespace: constants.Loki,
+			Name:      "ingest_limits_stream_evictions_total",
+			Help:      "The total number of streams evicted due to age per tenant. This is not a global total, as tenants can be sharded over multiple pods.",
+		}, []string{"tenant"}),
+		kafkaReadLatency: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
+			Namespace:                       constants.Loki,
+			Name:                            "ingest_limits_kafka_read_latency_seconds",
+			Help:                            "Latency to read stream metadata from Kafka.",
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMinResetDuration: 1 * time.Hour,
+			NativeHistogramMaxBucketNumber:  100,
+			Buckets:                         prometheus.DefBuckets,
+		}),
+		kafkaReadBytesTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace: constants.Loki,
+			Name:      "ingest_limits_kafka_read_bytes_total",
+			Help:      "Total number of bytes read from Kafka.",
+		}),
+	}
+}
+
+type streamMetadata struct {
+	hash       uint64
+	lastSeenAt int64
+}
+
+// IngestLimits is a service that manages stream metadata limits.
+type IngestLimits struct {
+	services.Service
+
+	cfg    Config
+	logger log.Logger
+	client *kgo.Client
+
+	lifecycler        *ring.Lifecycler
+	lifecyclerWatcher *services.FailureWatcher
+
+	// metrics
+	metrics *metrics
+
+	// Track stream metadata
+	mtx      sync.RWMutex
+	metadata map[string]map[int32][]streamMetadata // tenant -> partitionID -> streamMetadata
+
+	// Track partition assignments
+	mtxAssingedPartitions sync.RWMutex
```

Comment:
Need to remove this, we will just use `mtx`.

Reply:
I believe you have already removed that right?

Reply:
Yeah but it regressed somehow, need to remove it again.